src/lib/mod.rs

```diff
@@ -139,77 +140,77 @@ pub fn grab_pointer_set_cursor(conn: &xcb::Connection, root: u32) -> bool {
     false
 }
 
-pub fn find_escape_keycode(conn: &xcb::Connection) -> xcb::Keycode {
+pub fn find_escape_keycode<C: Connection>(conn: &C) -> xproto::Keycode {
     // https://stackoverflow.com/questions/18689863/obtain-keyboard-layout-and-keysyms-with-xcb
-    let setup = conn.get_setup();
-    let cookie = xcb::get_keyboard_mapping(
-        &conn,
-        setup.min_keycode(),
-        setup.max_keycode() - setup.min_keycode() + 1,
-    );
-    let reply = cookie.get_reply().expect("failed to get keyboard mapping");
+    let setup = conn.setup();
+    let cookie = xproto::get_keyboard_mapping(
+        conn,
+        setup.min_keycode,
+        setup.max_keycode - setup.min_keycode + 1,
+    ).unwrap();
+    let reply = cookie.reply().expect("failed to get keyboard mapping");
 
     let escape_index = reply
-        .keysyms()
+        .keysyms
         .iter()
         .position(|&keysym| keysym == ESC_KEYSYM)
         .expect("failed to find escape keysym");
-    (escape_index / reply.keysyms_per_keycode() as usize) as u8 + setup.min_keycode()
+    (escape_index / reply.keysyms_per_keycode as usize) as u8 + setup.min_keycode
 }
 
-pub fn grab_key(conn: &xcb::Connection, root: u32, keycode: u8) {
+pub fn grab_key<C: Connection>(conn: &C, root: u32, keycode: u8) {
     for mask in 0..=KEY_GRAB_MASK_MAX {
-        xcb::grab_key(
-            &conn,
+        xproto::grab_key(
+            conn,
             true,
             root,
-            mask as u16,
+            mask,
             keycode,
-            xcb::GRAB_MODE_ASYNC as u8,
-            xcb::GRAB_MODE_ASYNC as u8,
-        );
+            xproto::GrabMode::Async,
+            xproto::GrabMode::Async,
+        ).unwrap().check().unwrap();
     }
 }
 
-pub fn ungrab_key(conn: &xcb::Connection, root: u32, keycode: u8) {
+pub fn ungrab_key<C: Connection>(conn: &C, root: u32, keycode: u8) {
     for mask in 0..=KEY_GRAB_MASK_MAX {
-        xcb::ungrab_key(&conn, keycode, root, mask as u16);
+        xproto::ungrab_key(conn, keycode, root, mask).unwrap().check().unwrap();
     }
 }
 
-fn viewable(conn: &xcb::Connection, win: xcb::Window) -> bool {
-    let attrs = xcb::get_window_attributes(conn, win).get_reply().unwrap();
-    (attrs.map_state() & xcb::MAP_STATE_VIEWABLE as u8) != 0
+fn viewable<C: Connection>(conn: &C, win: xproto::Window) -> bool {
+    let attrs = xproto::get_window_attributes(conn, win).unwrap().reply().unwrap();
+    attrs.map_state == xproto::MapState::Viewable
```

Comment:
map_state is an enum variant so using == works instead of checking bits

Reply:
(this should probably go in a source code comment)

Reply:
It was an explanation for why the change is valid.